### ISSUE TITLE
Migrate to SSL by default

### ISSUE THREAD

**Issue by demianrenzulli** [ampproject/amp-publisher-sample#117]
The shadow reader works on HTTPS, but not by default.

Users that type "amp.cards" on the browser will land into a non-secure site, which won't register the SW (which is one of the main features of this demo).

**Comment by morsssss** [comment]
I'd call this a P1!  Do you have time to look into how to set SSL up on GCP?

https://cloud.google.com/sql/docs/mysql/configure-ssl-instance

@pbakaus does anything else need to be done to redirect to HTTPS?

**Comment by demianrenzulli** [comment]
@morsssss will await @pbakaus  comments and then start working on this. Will reach out to you to ask for access to the backend.

**Comment by pbakaus** [comment]
nothing else AFAIK, I just don't know how to properly set it up on GCP.

**Comment by demianrenzulli** [comment]
Hi guys,

I'm not seeing any reference to the domain "amp.cards" in the code itself, or any configuration file, outside the [data-iframe-src attribute](https://github.com/ampproject/amp-publisher-sample/blob/82185754b2f6b3b135d9fbb695a3d5f238e5db8b/amp-pwa-reader/src/server/partials/sw_body.html#L3) of the amp-install-serviceworker component.
With that said, I think that all the configuration should be done on GCP console.

[This](https://cloud.google.com/appengine/docs/standard/nodejs/securing-custom-domains-with-ssl) seems to be the guide to migrate to a Node Application to SSL when using custom domains.

If we can take a look at the GCP console, maybe we can check if all the required steps are correct.

**Comment by pbakaus** [comment]
I just looked into this for a little bit, and don't think there's anything to be done in GCP. Rather, it's HSTS (Strict Transport Security) headers that are missing, and I figured the obvious way to add them would be in the actual server: https://github.com/ampproject/amp-publisher-sample/blob/master/amp-pwa-reader/src/server/server.js

However, I'm at a loss why all of this even works on HTTPS today anyway, as the express server spun up in that file doesn't configure HTTPS...I wonder if GCP does wicked magic?

**Comment by demianrenzulli** [comment]
Thanks Paul, would you recommend to add the header directly on the code on server.js, or using something like [helmet](https://www.npmjs.com/package/helmet), that would encapsulate many security features (including HSTS)?

If the latter, the work would consist on adding one entry to package.json, and two lines of code to server.js.

**Comment by andreban** [comment]
Hey everyone. We use this config in PWA Directory, to make all URLs be served from a secure URL: https://github.com/GoogleChromeLabs/gulliver/blob/master/app.yaml#L18-L21, along with this bit: https://github.com/GoogleChromeLabs/gulliver/blob/master/app.js#L41-L49 Does this help?

**Comment by pbakaus** [comment]
oooh this looks good. Andre or Demian, are you cool providing a PR with this change?

**Comment by demianrenzulli** [comment]
This is what I tried adding locally to add helmet, but haven't tested in a live environment yet, to see if it actually works:

npm install helmet --save (will add helmet dependency to gulpfile).

On server.js:

const helmet = require('helmet')
app.use(helmet());

Before moving forward with this option, do you guys think that using [helmet](https://www.npmjs.com/package/helmet) is the way to go, or it's better to stick to Andre's solution (which seems to solve the particular SSL need, without adding anything else that might potentially be unnecessary)?

**Comment by demianrenzulli** [comment]
BTW: helmet is recommended as best practice on [ExpressJS site](https://expressjs.com/en/advanced/best-practice-security.html#use-helmet), but I don't have any other experience with it.

**Comment by pbakaus** [comment]
So the files Andre pointed out, the change in app.yaml is probably the most important as it relates to GCP.

The thing that I'm wondering about: Why does HTTPS even work today (as there is no code at all suggesting express is configured for it)? So anyway, I'm all for trying out helmet and doing the app.yaml change and see how it goes.

**Comment by demianrenzulli** [comment]
@pbakaus @morsssss @andreban created PR #119 and uploaded a version on GCP at: [steel-topic-208022.appspot.com](http://steel-topic-208022.appspot.com).

**Comment by morsssss** [comment]
Deployed! All looks good 👍

**Comment by demianrenzulli** [comment]
BTW, just cloned the latest version and uploaded it here: nodejs-tests-210101.appspot.com.
On my tests it is always redirecting to HTTPS, regardless of the protocol on the URL.

**Comment by morsssss** [comment]
That's funny.  For me, amp.cards did always redirect to HTTPS.  But if I
try that in an incognito window, it doesn't.

Your appspot version works for me in the incognito window, though.

When I deployed, all your new code was in the directory I deployed from.
Just confirmed that.  So this is a bit of a mystery...

I'm sure there's a way to actually check the code that's on the server.  If
I get a minute tomorrow I'll do that, but I may not get that minute....

**Comment by demianrenzulli** [comment]
Hi Ben! In case you test the url I sent before: I turned off nodejs-tests-210101.appspot.com, because I found I was starting to get charged again.
I can't explain the reasons since that version should be receiving no traffic (or minimal). Maybe there's some special configuration requirement on this project reserving resources.
Don't worry about checking the code, maybe we can take a look on Thursday at the production server to understand why it's not redirecting by default, or any other day.
